src/defs.h

```diff
@@ -536,53 +536,43 @@ namespace std {
 namespace tr1 {
 #endif
-template <> struct hash<struct sockaddr_in> : public std::unary_function<struct sockaddr_in, int> {
-    int operator()(struct sockaddr_in const &key) const {
-        // XOR "a.b" part of "a.b.c.d" address with 16bit port; leave "c.d" part untouched for
-        // maximum hashing
-        return key.sin_addr.s_addr ^ key.sin_port;
-    }
-};
-
-template <>
-struct hash<struct port_and_type> : public std::unary_function<struct port_and_type, int> {
-    int operator()(struct port_and_type const &key) const {
+template <> struct hash<struct sockaddr_store_t> : public std::unary_function<struct sockaddr_store_t, int> {
+    int operator()(struct sockaddr_store_t const &key) const {
         // XOR "a.b" part of "a.b.c.d" address with 16bit port; leave "c.d" part untouched for
         // maximum hashing
-        return key.sock_type ^ key.port;
+        // use only IPv4 part for hashing
+        return ((sockaddr_in &)key).sin_addr.s_addr ^ ((sockaddr_in &)key).sin_port;
     }
 };
-
-template <> struct hash<struct in_addr> : public std::unary_function<struct in_addr, int> {
-    int operator()(struct in_addr const &key) const { return key.s_addr & 0xFF; }
-};
 #if !defined(WIN32) && !defined(__FreeBSD__)
 } // closes namespace tr1
 #endif
 template <>
-struct equal_to<struct sockaddr_in> : public std::binary_function<struct sockaddr_in,
-                                                                  struct sockaddr_in, bool> {
-    bool operator()(struct sockaddr_in const &key1, struct sockaddr_in const &key2) const {
-        return key1.sin_port == key2.sin_port && key1.sin_addr.s_addr == key2.sin_addr.s_addr;
-    }
-};
-
-template <>
-struct equal_to<struct port_and_type> : public std::binary_function<struct port_and_type,
-                                                                    struct port_and_type, bool> {
-    bool operator()(struct port_and_type const &key1, struct port_and_type const &key2) const {
-        return key1.sock_type == key2.sock_type && key1.port == key2.port;
+struct equal_to<struct sockaddr_store_t> :
+        public std::binary_function<struct sockaddr_store_t,
+                struct sockaddr_store_t, bool> {
+    bool operator()(struct sockaddr_store_t const &key1, struct sockaddr_store_t const &key2) const {
+        if (key1.ss_family != key2.ss_family) {
+            return false;
+        }
+        switch (key1.ss_family) {
+        case AF_INET: {
+            const sockaddr_in &k1 = (const sockaddr_in &)key1;
+            const sockaddr_in &k2 = (const sockaddr_in &)key2;
+            return k1.sin_port == k2.sin_port &&
+                k1.sin_addr.s_addr == k2.sin_addr.s_addr;
+        }
+        case AF_INET6: {
+            const sockaddr_in6 &k1 = (const sockaddr_in6 &)key1;
+            const sockaddr_in6 &k2 = (const sockaddr_in6 &)key2;
+            return k1.sin6_port == k2.sin6_port &&
+                IN6_ARE_ADDR_EQUAL(&k1.sin6_addr, &k2.sin6_addr);
```

Comment:
Many impls do this with four 32bit comparisions. We can do it with two 64bit comparisions.

Reply:
Are you sure that data is always 8-byte aligned?

Reply:
I can't imagine how to express it in code without casting a casting a sin6_addr pointer to uint64_t pointer.

Reply:
I believe we can cast..

Reply:
Imagine that &addr == 0x100004. This is an aligned address for 32-bit integer `__in6_u.__u6_addr32[0]` but for 64-bit integer `__in6_u.__u6_addr64[0]` this address is not aligned. This will result in SIGBUS on SPARC and slower access on another architectures.